lib/ansible/plugins/action/bigip.py

```diff
@@ -0,0 +1,110 @@
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+#
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import sys
+import copy
+
+from ansible import constants as C
+from ansible.module_utils.basic import AnsibleFallbackNotFound
+from ansible.module_utils.f5_utils import F5_COMMON_ARGS
+from ansible.module_utils.six import iteritems
+from ansible.plugins.action.normal import ActionModule as _ActionModule
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+
+class ActionModule(_ActionModule):
+
+    def run(self, tmp=None, task_vars=None):
+        if self._play_context.connection != 'local':
+            return dict(
+                failed=True,
+                msg='invalid connection specified, expected connection=local, got %s' % self._play_context.connection
+            )
+
+        transport = self._task.args.get('transport', 'rest')
+
+        display.vvvv('connection transport is %s' % transport, self._play_context.remote_addr)
+
+        if transport == 'cli':
+            provider = self.load_provider()
+            pc = copy.deepcopy(self._play_context)
+            pc.connection = 'network_cli'
+            pc.network_os = 'bigip'
+            pc.remote_addr = provider.get('server', self._play_context.remote_addr)
+            pc.port = int(provider['server_port'] or self._play_context.port or 22)
+            pc.remote_user = provider.get('user', self._play_context.connection_user)
+            pc.password = provider.get('password', self._play_context.password)
+            pc.timeout = int(provider.get('timeout', C.PERSISTENT_COMMAND_TIMEOUT))
+
+            display.vvv('using connection plugin %s' % pc.connection, pc.remote_addr)
+            connection = self._shared_loader_obj.connection_loader.get('persistent', pc, sys.stdin)
+
+            socket_path = connection.run()
+            display.vvvv('socket_path: %s' % socket_path, pc.remote_addr)
+            if not socket_path:
+                return {'failed': True,
+                        'msg': 'unable to open shell. Please see: ' +
+                               'https://docs.ansible.com/ansible/network_debug_troubleshooting.html#unable-to-open-shell'}
+
+            # make sure we are in the right cli context which should be
+            # enable mode and not config mode
+            rc, out, err = connection.exec_command('prompt()')
+            while '(config' in str(out):
+                display.vvvv('wrong context, sending exit to device', self._play_context.remote_addr)
+                connection.exec_command('exit')
+                rc, out, err = connection.exec_command('prompt()')
+
+            task_vars['ansible_socket'] = socket_path
+
+        result = super(ActionModule, self).run(tmp, task_vars)
+        return result
+
+    def load_provider(self):
```

Comment:
In 2.5 the logic to load provider arguments is moved at a common place in [module_utils/network_common.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ios.py#L48)
Please refer [here](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/action/ios.py#L48) for reference.

Reply:
I don't believe we can do that since the provider isn't in the default args. The one in network_common adds the provider `args['provider'] = provider` which will result in the other F5 modules erroring with `Unsupported parameters for (bigip_config) module: provider`

Reply:
I see the issue, in this case, `F5_COMMON_ARGS` is similar to `provider`  but not required to be updated in task args. For other platforms `provider` need to be part of task args as provider spec has some default values which are directly referenced in the module code. 

One way to avoid duplicating this code is to delete `provider` from task args in `bigip` action plugin before calling `run()`.

If this is applicable for other platforms in future I think it is better to move the line that assigns `provider` to task args in each of the action plugins. Thoughts?

Reply:
I'll let @caphrim007 decide since he's the one maintaining all the F5 modules.
I'm ok with deleting `provider` from the task args in the `bigip` action plugin. Always like reusing code vs duplicating it.

Reply:
deleting and re-using the existing code is fine.